src/policy/space.rs

```diff
@@ -67,14 +67,16 @@ pub trait Space<VM: VMBinding>: 'static + SFT + Sync + Downcast {
     /// avoid arithmatic overflow. If we have to do computation in the allocation fastpath and
     /// overflow happens there, there is nothing we can do about it.
     /// Return a boolean to indicate if we will be out of memory, determined by the check.
-    fn will_oom_on_acquire(&self, tls: VMThread, size: usize) -> bool {
+    fn will_oom_on_acquire(&self, tls: VMThread, size: usize, allow_oom_call: bool) -> bool {
```

Comment:
I suggest splitting `will_oom_on_acquire` into two functions.

-   `will_oom_on_acquire`: This returns `bool`.  If it would OOM, it returns `true`, but doesn't actually call `out_of_memory`.
-   `handle_obvious_oom_request`: Call `will_oom_on_acquire` and, if it returns true, call `out_of_memory`.

One reason is that the name "will_..." suggests it only performs a check, and will not have side effects.  Another reason for this change is that there is a call site in `Space::acquire`:

```rust
        debug_assert!(
            !self.will_oom_on_acquire(tls, pages << LOG_BYTES_IN_PAGE, !no_gc_on_fail),
            "The requested pages is larger than the max heap size. Is will_go_oom_on_acquire used before acquring memory?"
        );
```

This assertion will never fail, but will panic in `Collection::out_of_memory`.  After splitting, `will_oom_on_acquire` will not call `out_of_memory`, and the assertion may fail if it returns false.

There were two more call sites of `will_oom_on_acquire`, in `bumpallocator.rs` and `large_object_allocator.rs`, respectively.  After the split, those two call sites shall call `handle_obvious_oom_request` instead.

I drafted a commit so you can cherry-pick: https://github.com/mmtk/mmtk-core/commit/b6da92ef89389ee0ecbf88ea7810c31e2ec3f2cd

Reply:
Sounds good. I cherry-picked your changes.